src/Service/HourReportService.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+namespace App\Service;
+
+use App\Exception\EconomicsException;
+use App\Model\Reports\HourReportData;
+use App\Model\Reports\HourReportProjectTag;
+use App\Model\Reports\HourReportProjectTicket;
+use App\Model\Reports\HourReportTimesheet;
+use App\Repository\IssueRepository;
+use App\Repository\ProjectRepository;
+use App\Repository\ProjectVersionBudgetRepository;
+use App\Repository\VersionRepository;
+use App\Repository\WorklogRepository;
+use Doctrine\ORM\EntityManagerInterface;
+
+class HourReportService
+{
+    public function __construct(
+        private readonly ProjectVersionBudgetRepository $budgetRepository,
+        private readonly EntityManagerInterface $entityManager,
+        private readonly IssueRepository $issueRepository,
+        private readonly WorklogRepository $worklogRepository,
+        private readonly ProjectRepository $projectRepository,
+        private readonly VersionRepository $versionRepository,
+    ) {
+    }
+
+    /**
+     * @throws EconomicsException
+     */
+    public function getProjects(): array
+    {
+        $projects = $this->projectRepository->findAll();
+        $projectChoices = [];
+        foreach ($projects as $project) {
+            $projectChoices[$project->getName()] = $project->getId();
+        }
+
+        return $projectChoices;
+    }
+
+    /**
+     * @throws EconomicsException
+     */
+    public function getMilestones(string $projectId, bool $allAllOption = false): array
+    {
+        $milestones = $this->versionRepository->findBy(['project' => $projectId]);
+        $milestoneChoices = [];
+        if ($allAllOption) {
+            $milestoneChoices['All milestones'] = '0';
+        }
+        foreach ($milestones as $milestone) {
+            $milestoneName = $milestone->getName() ?? '';
+            $milestoneChoices[$milestoneName] = $milestone->getId();
+        }
+
+        return $milestoneChoices;
+    }
+
+    /**
+     * @throws EconomicsException
+     */
+    public function getHourReport(string $projectId, ?\DateTimeInterface $fromDate, ?\DateTimeInterface $toDate, int $versionId = null): HourReportData
+    {
+        if (!$projectId) {
+            throw new EconomicsException('No project id specified');
+        }
+        $hourReportData = new HourReportData(0, 0);
+        $projectIssues = $this->issueRepository->findBy(['project' => $projectId]);
+
+        foreach ($projectIssues as $issue) {
+            // If version is provided, we only want the issues containing the versionId
+            if ($versionId) {
+                $issueHasVersion = $issue->getVersions()->exists(function ($key, $value) use ($versionId) {
```

Comment:
Instead of iterating all issues for the project, when issueId is provided, you could use
```
public function issueContainingVersion(int $versionId): array
    {
        $qb = $this->createQueryBuilder('issue')
            ->where(':version MEMBER OF issue.versions')
            ->setParameter('version', $versionId);

        return $qb->getQuery()->getResult();
    }
```

Replace:
`$projectIssues = $this->issueRepository->findBy(['project' => $projectId]);`
with

```
if ($versionId) {
  $this->issueRepository->issueContainingVersion($versionId);
} else {
  $projectIssues = $this->issueRepository->findBy(['project' => $projectId]);
}
```

And then remove if($versionId) block. This would save a lot of computations, because you only iterate what is needed.